akka-stream-tests/src/test/scala/akka/stream/io/compression/CoderSpec.scala

```diff
@@ -0,0 +1,194 @@
+      val compressed = comp.compressAndFinish(inputBytes)
+
+      ourDecode(compressed) should equal(inputBytes)
+    }
+
+    "shouldn't produce huge ByteStrings for some input" in {
+      val array = new Array[Byte](10) // FIXME
+      util.Arrays.fill(array, 1.toByte)
+      val compressed = streamEncode(ByteString(array))
+      val limit = 10000
+      val resultBs =
+        Source.single(compressed)
+          .via(decoderFlow(maxBytesPerChunk = limit))
+          .limit(4200).runWith(Sink.seq)
+          .awaitResult(3.seconds)
+
+      forAll(resultBs) { bs ⇒
+        bs.length should be < limit
+        bs.forall(_ == 1) should equal(true)
+      }
+    }
+
+    "be able to decode chunk-by-chunk (depending on input chunks)" in {
+      val minLength = 100
+      val maxLength = 1000
+      val numElements = 1000
+
+      val random = ThreadLocalRandom.current()
+      val sizes = Seq.fill(numElements)(random.nextInt(minLength, maxLength))
+      def createByteString(size: Int): ByteString =
+        ByteString(Array.fill(size)(1.toByte))
+
+      val sizesAfterRoundtrip =
+        Source.fromIterator(() ⇒ sizes.toIterator.map(createByteString))
+          .via(encoderFlow)
+          .via(decoderFlow())
+          .runFold(Seq.empty[Int])(_ :+ _.size)
+
+      sizesAfterRoundtrip.awaitResult(3.seconds) shouldEqual sizes
+    }
+
+    extraTests()
+  }
+
+  def encode(s: String) = ourEncode(ByteString(s, "UTF8"))
+  def ourEncode(bytes: ByteString): ByteString = newCompressor().compressAndFinish(bytes)
+  def ourDecode(bytes: ByteString): ByteString =
+    Source.single(bytes)
+      .via(decoderFlow())
+      .join
+      .awaitResult(3.seconds)
+
+  lazy val corruptContent = {
+    val content = encode(largeText).toArray
+    content(14) = 26.toByte
+    ByteString(content)
+  }
+
+  def streamEncode(bytes: ByteString): ByteString = {
```

Comment:
I know it is  `{Input|Output}Stream` but this confused me a bit where it's used, contra akka streams, `blockingEncode`, `imperativeEncode` or such would have helped.